js/dataframe/cell.tsx

```diff
@@ -0,0 +1,279 @@
+    }
+  }, [editRowIndex, editColumnIndex, rowIndex, columnIndex]);
+
+  const hasUpdated = cellEditMap.has(`[${rowIndex}, ${columnIndex}]`);
+
+  const tableCellClass = CellStateClassMap[cellState];
+
+  const [errorTitle, setErrorTitle] = useState<string | undefined>(undefined);
+
+  const resetEditInfo = React.useCallback(() => {
+    setEditRowIndex(null);
+    setEditColumnIndex(null);
+  }, [setEditRowIndex, setEditColumnIndex]);
+
+  const onEsc = (e: ReactKeyboardEvent<HTMLInputElement>) => {
+    if (e.key !== "Escape") return;
+    // Prevent default behavior
+    e.preventDefault();
+
+    // inputRef.current.blur();
+    resetEditInfo();
+    // TODO-barret-future; Set focus to table? (state: Editing was aborted)
+  };
+  const onTab = (e: ReactKeyboardEvent<HTMLInputElement>) => {
+    if (e.key !== "Tab") return;
+    // Prevent default behavior
+    e.preventDefault();
+
+    const hasShift = e.shiftKey;
+
+    const newColumnIndex = editColumnIndex! + (hasShift ? -1 : 1);
+    if (newColumnIndex < 0 || newColumnIndex >= columns.length) {
+      // If the new column index is out of bounds, quit
+      return;
+    }
+
+    setEditColumnIndex(newColumnIndex);
+  };
+  const onEnter = (e: ReactKeyboardEvent<HTMLInputElement>) => {
+    if (e.key !== "Enter") return;
+    // Prevent default behavior
+    e.preventDefault();
+
+    const hasShift = e.shiftKey;
+
+    const newRowIndex = editRowIndex! + (hasShift ? -1 : 1);
+    if (newRowIndex < 0 || newRowIndex >= maxRowSize) {
+      // If the new row index is out of bounds, quit
+      // attemptUpdate();
+      return;
+    }
+
+    setEditRowIndex(newRowIndex);
+  };
+
+  const onInputKeyDown = (e: ReactKeyboardEvent<HTMLInputElement>) => {
+    [onEsc, onEnter, onTab].forEach((fn) => fn(e));
+  };
+
+  const attemptUpdate = () => {
+    // Only update if the string form of the value has changed
+    if (`${initialValue}` === `${value}`) return;
+
+    setCellState(CellStateMap.EditSaving);
+    // console.log("Setting update count");
+    // setUpdateCount(updateCount + 1);
+    updateCellsData({
+      id,
+      cellInfos: [
+        {
+          rowIndex,
+          columnIndex,
+          value,
+          prev: initialValue,
+        },
+      ],
+      onSuccess: (values) => {
+        // console.log("success!", values);
+        setCellState(CellStateMap.EditSuccess);
+      },
+      onError: (err) => {
+        console.error("Error saving tabel cell value!", err);
+        setErrorTitle(String(err));
+        setCellState(CellStateMap.EditFailure);
+      },
+      columns,
+      setData,
+      setCellEditMap,
+    });
+  };
+
+  // When the input is blurred, we'll call our table meta's updateData function
+  // console.log("rendering cell", rowIndex, id, initialValue, value);
+  const onBlur = () => {
+    // console.log("on blur!", initialValue, value, e);
+    attemptUpdate();
+  };
+
+  // If the initialValue is changed external, sync it up with our state
```

Comment:
```suggestion
  // If the initialValue is changed externally, sync it up with our state
```

Reply:
Also say something about how this external change can happen (not from editing the cell, but from copy/paste)